### PR TITLE
doctest for .rst files

### PR DESCRIPTION
**Pull request by awaelchli**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Follow up to #1484 in a series of doc  improvements.
This PR adds doctesting for .rst files and converts existing code blocks to proper doctests. 
**Note:** 

Notes: 
- For the tests to work I had to change imports like `pl.LightningModule` to because otherwise there was an import error because in docs the `__LIGHTNING_SETUP__` variable is set.
- Even the doctests that don't test for outputs (like class defs) are useful because it still checks syntax. In fact, while converting them I found syntax errors that would otherwise not have been detected. So if someone copy pastes from docs, it is less likely they need to fix a typo we introduced.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
pls add also testing RST files with pytest along with the existing testing because now it is tested only against one package's configuration but we would like to cover at least min and max package version coverage >> add it to Github action CI

**Comment by awaelchli** [comment]
this is just a draft and I'm experimenting. I don't know how to do all these things but  I will figure it out. 

P.S. you deleted all my content in the PR description. can it be undone? I have to type it again :(

**Comment by Borda** [comment]
> P.S. you deleted all my content in the PR description. can it be undone? I have to type it again :(

I didn't, you can see the version history as next to the content title is edited by ...

**Comment by awaelchli** [comment]
@Borda The doctestplus you suggested seems not to work very well with sphinx. It is a library to run doctests on standalone rst files with pytest. However, since we are using rst files with sphinx, doctestplus is not optional. First of all, sphinx doctests support things like this:
```
.. doctest::
    :skipif: torch.cuda.device_count() < 2
    >>> Trainer(gpus=2)
```
But this is not supported by [doctestplus](https://github.com/astropy/pytest-doctestplus) and it would be nice to run our doctests conditionally. 
So, I suggest to use simply the sphinx built in doctest. 
What do you think?

**Comment by Borda** [comment]
the point is that the Read-The-docs is running the build on quite a simple machine so we do not want to take it too long... let's try it with sphinx only and see how it goes :]

File: docs/source/apex.rst [hunk 1]
```diff
@@ -41,7 +46,7 @@ Enable 16-bit
 .. code-block:: python
 
     # turn on 16-bit
-    trainer = Trainer(amp_level='O1', precision=16)
+    >>> trainer = Trainer(amp_level='O1', precision=16)  # doctest: +SKIP
```
Comment (Borda) [inline docs/source/apex.rst:49]: if is it skipped there is no need to test it, but why skip this?

Comment (awaelchli) [reply docs/source/apex.rst:49]: yes I agree. If I turn it back to sphinx doctest we can explicitly skip only if the machine does not have amp. I have the sphinx stuff on a different branch, I will force push it here then I can finishe it for review, ok?

**Comment by Borda** [comment]
> The [doctestplus](https://github.com/astropy/pytest-doctestplus) you suggested seems not to work very well with sphinx

what does it mean, some tests are skipped or failing?

**Comment by awaelchli** [comment]
No it means it is not designed to be used together with sphinx. It will ignore directives like 
```
.. doctest::
    :skipif: torch.cuda.device_count() < 2
    >>> Trainer(gpus=2)
```
Sphinx has its own doctest extension.

**Comment by williamFalcon** [comment]
@awaelchli @Borda shall we merge soon? unless @awaelchli wants to keep rebasing lol... this will get hard to merge the longer it's open :)

**Comment by awaelchli** [comment]
@williamFalcon sure, I'll try to finish it today, thanks.

**Comment by awaelchli** [comment]
@Borda in the doctests I import loggers, should I add requirements-extra to the docs build or what should I do?

**Comment by Borda** [comment]
> @Borda in the doctests I import loggers, should I add requirements-extra to the docs build or what should I do?

in this case, I think yeas you really need to call some methods... the mock here is not enough, right?

**Comment by awaelchli** [comment]
hmm, not sure how I could mock them. 
It seems not possible to just simply install requirements-extra because of horovod. 
Should I remove doctesting for loggers completely?

**Comment by Borda** [comment]
> hmm, not sure how I could mock them.

I think it is not possible if we want to run the tests...

> It seems not possible to just simply install requirements-extra because of horovod.

what is the problem with Horovod?

**Comment by awaelchli** [comment]
It seems CI can't install it if I do pip install -r requirements-extra.txt

**Comment by awaelchli** [comment]
@tgaddair Hi do you know how I can install or ignore horovod in the docs build CI? It is CPU only and I need to install the logger requiremetns in requirements-extra.txt but it also contains horovod.

**Comment by tgaddair** [comment]
> @tgaddair Hi do you know how I can install or ignore horovod in the docs build CI? It is CPU only and I need to install the logger requiremetns in requirements-extra.txt but it also contains horovod.

Hey @awaelchli, Horovod should be able to install with or without GPU.  I can take a quick look at the build output to see why it's failing.

**Comment by awaelchli** [comment]
Thanks, here it is:
https://circleci.com/gh/PyTorchLightning/pytorch-lightning/25540?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
Ignore the other failures, they are not due to horovod.

**Comment by tgaddair** [comment]
> Thanks, here it is:
> https://circleci.com/gh/PyTorchLightning/pytorch-lightning/25540?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
> Ignore the other failures, they are not due to horovod.

Thanks.  From the logs, it seems the problem is that neither MPI nor CMake are installed.  One of those most be installed for Horovod to build.  I suggest installing CMake as we do in `install_deps`:

```
sudo apt-get update && sudo apt-get install -y cmake
```

That should take care of the issue.

**Comment by awaelchli** [comment]
@tgaddair Thanks for the help, it worked. 
@Borda the drone has a red cross but when I look at the output, all tests are passing, no mention of a failing test. could you have a quick look and check if you see something obvious I don't see?

File: .drone.yml [hunk 2]
```diff
@@ -34,10 +34,13 @@ steps:
     #- bash ./tests/install_AMP.sh
     - apt-get update && apt-get install -y cmake
     - pip install -r requirements.txt --user -q
+    - pip install -r requirements-extra.txt --user -q
```
Comment (Borda) [inline .drone.yml:37]: this is already included in devel below

Comment (awaelchli) [reply .drone.yml]: Yep, thanks. 
There is still an issue with drone. It does not give a  green mark even though all tests pass. Somehow sphinx doctest is not sending a success signal.

Comment (awaelchli) [reply .drone.yml]: I will try with a newer sphinx version

**Comment by awaelchli** [comment]
Can't invest more time into trying to make the doctests run on drone. I have have tried anything, there is just no output that would suggest any failures. So I suggest to drop it for drone and just keep it in the circle ci. 

PR is now ready for review

**Comment by Borda** [comment]
> Can't invest more time into trying to make the doctests run on drone. I have tried everything, there is just no output that would suggest any failures. So I suggest to drop it for drone and just keep it in the circle ci.

I am fine with having it now at least for CPU (CircleCI) and GPU we can add later...

File: docs/source/conf.py [hunk 3]
```diff
@@ -316,19 +316,10 @@ def setup(app):
 
 # TODO: better parse from package since the import name and package name may differ
 MOCK_MANUAL_PACKAGES = [
-    'torch',
     'torchvision',
```
Comment (Borda) [inline docs/source/conf.py:319]: this is now also in extras

File: .drone.yml [hunk 4]
```diff
@@ -35,9 +35,11 @@ steps:
     - apt-get update && apt-get install -y cmake
     - pip install -r requirements.txt --user -q
     - pip install -r ./tests/requirements-devel.txt --user -q
+    - pip install -r ./docs/requirements.txt --user -q
```
Comment (Borda) [inline .drone.yml:38]:
```suggestion
    #- pip install -r ./docs/requirements.txt --user -q
```

File: docs/source/conf.py [hunk 5]
```diff
@@ -307,7 +307,7 @@ def setup(app):
 # https://stackoverflow.com/questions/15889621/sphinx-how-to-exclude-imports-in-automodule
 
 MOCK_REQUIRE_PACKAGES = []
-with open(os.path.join(PATH_ROOT, 'requirements.txt'), 'r') as fp:
+with open(os.path.join(PATH_ROOT, 'requirements-extra.txt'), 'r') as fp:
```
Comment (Borda) [inline docs/source/conf.py:310]: can split running the test and making docs, I mean that for just build doc via ReadTheDocs we would use just mock and for other CI like CircleCI we would do the full import...
I guess that there as to be a flag which distinguishes between `make html` and `make tests`

File: requirements-extra.txt [hunk 6]
```diff
@@ -7,5 +7,6 @@ test_tube>=0.7.5
 wandb>=0.8.21
 trains>=0.14.1
 matplotlib>=3.1.1
+torchvision>=0.4.1
```
Comment (Borda) [inline requirements-extra.txt:10]: we have removed it bc it was quite delayed on py3.8 so let's see if they have it there now...

Comment (awaelchli) [reply requirements-extra.txt:10]: You are right, just tested installing 0.4.1 in a py3.8 env and it failed. 
Kinda need it for some doctests. Two options:
1. ignore some of the doctests
2. manually install torchvision only in build docs CI task
what do you think is best?

File: docs/source/conf.py (hunk 5 above)
Comment (awaelchli) [reply docs/source/conf.py:310]: there is a RTD env variable called `READTHEDOCS` we could check
https://docs.readthedocs.io/en/stable/builds.html

File: requirements-extra.txt (hunk 6 above)
Comment (Borda) [reply requirements-extra.txt:10]: depends on how many tests need it... I would see a bit different option
- ignore the test with `torchvision` if they are just a few
- try to replace `torchvision` by our fund (probably too much work)
- run these tests only on CicleCI as we use there only 3.6/3.7

Comment (Borda) [reply requirements-extra.txt:10]: probably the last is the simplest way even we sacrify wider test cases test coverage

Comment (Borda) [reply requirements-extra.txt:10]: it seems that we are not testing the doc build in GitHub action anyway now so seems fast the less painful way :]

File: docs/source/conf.py (hunk 5 above)
Comment (Borda) [reply docs/source/conf.py:310]: ok, on the other hand, the RTD docs if build only with push to master which is not so often
so let's keep it as it is now and if we find that the build-time is an issue we'll change it

Comment (awaelchli) [reply docs/source/conf.py:310]: ok good, in case we need it I can do it quickly

File: requirements-extra.txt (hunk 6 above)
Comment (awaelchli) [reply requirements-extra.txt]: It was actually only 5 to skip. I added a check if torchvision is available and skip the tests. this way we can still tests this code on systems where torchvision is installed.

**Comment by awaelchli** [comment]
Looks like ReadTheDocs did not like what we modified in the conf.py. Some pages like Trainer and LightningModule are broken. I suspect it's the mocking of packages. Will track this down and submit a follow up PR.

**Comment by Borda** [comment]
@awaelchli could pls prepare fix as main repo branch and I will turn on building this branch so we will constantly know how it is going...

**Pull request merged**